app/questionnaire/router.py

```diff
@@ -55,13 +58,35 @@ def _get_survey_redirect_location(self):
             return None
 
         expected_location = self._get_first_incomplete_location_in_survey()
-        block = self._schema.get_block(self._current_location.block_id)
+        block = self._schema.block(self._current_location.block_id)
         cannot_access_survey_end = self._current_location != expected_location
 
         if block['type'] in ['Confirmation', 'Summary'] and cannot_access_survey_end:
             return expected_location
 
     def _get_first_incomplete_location_in_survey(self):
-        for location in self._routing_path:
-            if location not in self._completed_locations:
-                return location
+        sections = [section['id'] for section in self._schema.sections()]
+        completed_sections = [section for section in self._completed_locations.sections]
+        incomplete_sections = [
+            section for section in sections if section not in completed_sections
+        ]
+
+        if not incomplete_sections:
+            return
+
+        first_incomplete_section = self._schema.section(incomplete_sections[0])
+        completed_blocks = [
+            block.block_id for block in self._completed_locations.blocks
+        ]
+
+        incomplete_blocks = [
+            block
+            for group in first_incomplete_section['groups']
+            for block in group['blocks']
+            if block['id'] not in completed_blocks
+        ]
```

Comment:
This will loop over all blocks in a group regardless if they're complete or not. As we only care about the first incomplete block it would be better to return as soon as we get one.